frontend/src/utils/document.ts

```diff
@@ -0,0 +1,84 @@
+import { useEffect, useState, useCallback } from 'react';
+import { Editor } from 'slate';
+import { Document } from '../editor/types';
+
+export function getSpeakerName(
+  speaker: string | null,
+  speaker_names: Record<string, string>,
+): string {
+  if (!speaker) {
+    return `Unknown Speaker`;
+  } else if (speaker in speaker_names) {
+    return speaker_names[speaker];
+  } else {
+    return `Unnamed Speaker ${speaker}`;
+  }
+}
+
+export function useDocument<T>(editor: Editor, callback: (doc: Document) => T) {
```

Comment:
can we call callback `selector` or `lens` ?

Reply:
also: why does this not use useSlate internally?

Reply:
It should use `useSlateStatic` internally, `useSlate` I am not convinced by, because using `onDocChange` we can avoid a lot of recalculations (because we only catch actual document changes, instead of changes to selection for example)